src/main/java/org/jabref/model/openoffice/style/Citation.java

```diff
@@ -0,0 +1,142 @@
+package org.jabref.model.openoffice.style;
+
+import java.util.List;
+import java.util.Optional;
+
+import org.jabref.model.database.BibDatabase;
+import org.jabref.model.entry.BibEntry;
+import org.jabref.model.openoffice.ootext.OOText;
+import org.jabref.model.openoffice.util.OOPair;
+
+public class Citation implements ComparableCitation, CitationMarkerEntry, CitationMarkerNumericEntry {
+
+    /** key in database */
+    public final String citationKey;
+
+    /** Result from database lookup. Optional.empty() if not found. */
+    private Optional<CitationLookupResult> db;
+
+    /** The number used for numbered citation styles . */
+    private Optional<Integer> number;
+
+    /** Letter that makes the in-text citation unique. */
+    private Optional<String> uniqueLetter;
+
+    /** pageInfo */
+    private Optional<OOText> pageInfo;
+
+    /** isFirstAppearanceOfSource */
+    private boolean isFirstAppearanceOfSource;
+
+    /**
+     *
+     */
+    public Citation(String citationKey) {
+        this.citationKey = citationKey;
+        this.db = Optional.empty();
+        this.number = Optional.empty();
+        this.uniqueLetter = Optional.empty();
+        this.pageInfo = Optional.empty();
+        this.isFirstAppearanceOfSource = false;
+    }
+
+    @Override
+    public String getCitationKey() {
+        return citationKey;
+    }
+
+    @Override
+    public Optional<OOText> getPageInfo() {
+        return pageInfo;
+    }
+
+    @Override
+    public boolean getIsFirstAppearanceOfSource() {
+        return isFirstAppearanceOfSource;
+    }
+
+    @Override
+    public Optional<BibEntry> getBibEntry() {
+        return (db.isPresent()
+                ? Optional.of(db.get().entry)
+                : Optional.empty());
+    }
+
+    public static Optional<CitationLookupResult> lookup(List<BibDatabase> databases, String key) {
+        for (BibDatabase database : databases) {
```

Comment:
something like this could work:
result =   databases.stream().findFirst(database -> database.getEntryByCitationKey())
result.flatMap(...)

Reply:
original
```java
        for (BibDatabase database : databases) {
            Optional<BibEntry> entry = database.getEntryByCitationKey(key);
            if (entry.isPresent()) {
                return Optional.of(new CitationLookupResult(entry.get(), database));
            }
        }
        return Optional.empty();
```

Need `database` in `new CitationLookupResult(entry.get(), database)`

I can express it as

```java
        return databases.stream()
            .map(database ->
                 database.getEntryByCitationKey(key)
                 .map(e -> new CitationLookupResult(e, database)))
            .filter(Optional::isPresent)
            .findFirst()
            .flatMap(e -> e);
```

but it looks more complicated than the original.  Did I miss some opportunity to simplify?

Reply:
The stream version is ok.
It's maybe a question of taste, I personally prefer the stream version, because I like the fluid interface. I think it can be read and understood quicker.

Reply:
I miss the types. Added lookup from a single database.
The nice thing with streams is that usually steps can be written in the order they are performed.
Here findFirst shortcuts two (or three in the version above) steps later.

Changed to
```java
    public static Optional<CitationLookupResult> lookup(BibDatabase database, String key) {
        return (database
                .getEntryByCitationKey(key)
                .map(bibEntry -> new CitationLookupResult(bibEntry, database)));
    }

    public static Optional<CitationLookupResult> lookup(List<BibDatabase> databases, String key) {
        return (databases.stream()
                .map(database -> Citation.lookup(database, key))
                .filter(Optional::isPresent)
                .findFirst()
                .orElse(Optional.empty()));
    }
```

Reply:
Changed again, to avoid Optional<Optional<T>>, Optional::get first

```java
    public static Optional<CitationLookupResult> lookup(List<BibDatabase> databases, String key) {
        return (databases.stream()
                .map(database -> Citation.lookup(database, key))
                .filter(Optional::isPresent)
                .map(Optional::get)
                .findFirst());
    }
```